models/repo_list.go

```diff
@@ -115,76 +115,109 @@ type SearchRepoOptions struct {
+	}
+
+	// Check if user with Owner ID exists
+	if opts.Searcher == nil || opts.Searcher.ID != opts.OwnerID {
+		userExists, err := GetUser(&User{ID: opts.OwnerID})
+		if err != nil {
+			return nil, 0, err
+		}
+		if userExists == false {
+			return nil, 0, fmt.Errorf("User with OwnerID doesn't exist")
+		}
+	}
+
+	// Set public search by default
+	isPublicSearch := true
+
+	// Allow to search for owners private data
+	if opts.Searcher != nil && (opts.Searcher.ID == opts.OwnerID || opts.Searcher.IsAdmin) {
+		isPublicSearch = false
+	}
+
+	// Check and set page to correct number
 	if opts.Page <= 0 {
 		opts.Page = 1
 	}
 
-	if opts.Starred && opts.OwnerID > 0 {
+	var cond = builder.NewCond()
+
+	// Include starred repositories by Owner
+	if opts.Starred {
 		cond = builder.Eq{
 			"star.uid": opts.OwnerID,
 		}
 	}
 
-	opts.Keyword = strings.ToLower(opts.Keyword)
+	// Add repository name keyword to search for
 	if opts.Keyword != "" {
+		opts.Keyword = strings.ToLower(opts.Keyword)
 		cond = cond.And(builder.Like{"lower_name", opts.Keyword})
 	}
 
-	// Append conditions
-	if !opts.Starred && opts.OwnerID > 0 {
-		cond = cond.And(builder.Eq{"owner_id": opts.OwnerID})
-	}
-	if !opts.Private {
+	// Exclude private repositories
+	// if it is set in options
+	// or is public search
+	if !opts.Private || isPublicSearch {
 		cond = cond.And(builder.Eq{"is_private": false})
 	}
 
-	if opts.Searcher != nil {
-		var ownerIds []int64
+	// Set user access conditions
+	var accessCond = builder.NewCond()
 
-		ownerIds = append(ownerIds, opts.Searcher.ID)
-		err = opts.Searcher.GetOrganizations(true)
+	// Add Owner ID to access conditions
+	accessCond = builder.Eq{"owner_id": opts.OwnerID}
+
+	// Include collaborative repositories
+	if opts.Collaborate {
+		// Load owner's organizations
+		orgs, err := GetOrgUsersByUserID(opts.OwnerID, !isPublicSearch)
 
 		if err != nil {
 			return nil, 0, fmt.Errorf("Organization: %v", err)
 		}
 
-		for _, org := range opts.Searcher.Orgs {
+		var ownerIds []int64
+		for _, org := range orgs {
 			ownerIds = append(ownerIds, org.ID)
 		}
 
-		searcherReposCond := builder.In("owner_id", ownerIds)
-		if opts.Collaborate {
-			searcherReposCond = searcherReposCond.Or(builder.Expr(`id IN (SELECT repo_id FROM "access" WHERE access.user_id = ? AND owner_id != ?)`,
-				opts.Searcher.ID, opts.Searcher.ID))
-		}
-		cond = cond.And(searcherReposCond)
+		// Add repositories from related organizations
+		accessCond = accessCond.Or(builder.In("owner_id", ownerIds))
+
+		// Add repositories where user is set as collaborator directly
+		accessCond = accessCond.Or(builder.Expr(`id IN (SELECT repo_id FROM "access" WHERE access.user_id = ? AND owner_id != ?)`,
+			opts.OwnerID, opts.OwnerID))
 	}
 
+	// Add user access conditions to search
+	cond = cond.And(accessCond)
+
 	if len(opts.OrderBy) == 0 {
 		opts.OrderBy = "name ASC"
 	}
 
 	sess := x.NewSession()
 	defer sess.Close()
 
-	if opts.Starred && opts.OwnerID > 0 {
```

Comment:
`&& opts.OwnerID > 0` is a duplicate of https://github.com/Morlinest/gitea/blob/f4729438343b349ea99caa38a4dfef4cefb107d5/models/repo_list.go#L124.